src/monodroid/jni/embedded-assemblies-zip.cc

```diff
@@ -2,79 +2,131 @@
 		return false;
 	}
 
-	if (entry_name.get ()[0] != state.prefix[0] || strncmp (state.prefix, entry_name.get (), state.prefix_len) != 0) {
-		return false;
+	if (entry_name.get ()[0] != state.prefix[0] || memcmp (state.prefix, entry_name.get (), state.prefix_len) != 0) {
+		if (state.prefix == apk_lib_prefix.data ()) {
+			return false;
+		}
+
+		if (entry_name.get ()[0] != apk_lib_prefix[0] || memcmp (apk_lib_prefix.data (), entry_name.get (), apk_lib_prefix.size () - 1) != 0) {
+			return false;
+		}
 	}
 
 	if (application_config.have_runtime_config_blob && !runtime_config_blob_found) {
 		if (utils.ends_with (entry_name, SharedConstants::RUNTIME_CONFIG_BLOB_NAME)) {
 			runtime_config_blob_found = true;
-			runtime_config_blob_mmap = md_mmap_apk_file (state.apk_fd, state.data_offset, state.file_size, entry_name.get ());
+			runtime_config_blob_mmap = md_mmap_apk_file (state.file_fd, state.data_offset, state.file_size, entry_name.get ());
 			return false;
 		}
 	}
 
 	// assemblies must be 4-byte aligned, or Bad Things happen
 	if ((state.data_offset & 0x3) != 0) {
 		log_fatal (LOG_ASSEMBLY, "Assembly '%s' is located at bad offset %lu within the .apk\n", entry_name.get (), state.data_offset);
-		log_fatal (LOG_ASSEMBLY, "You MUST run `zipalign` on %s\n", strrchr (state.apk_name, '/') + 1);
+		log_fatal (LOG_ASSEMBLY, "You MUST run `zipalign` on %s\n", strrchr (state.file_name, '/') + 1);
 		Helpers::abort_application ();
 	}
 
 	return true;
 }
 
+inline void
+EmbeddedAssemblies::store_individual_assembly_data (dynamic_local_string<SENSIBLE_PATH_MAX> const& entry_name, ZipEntryLoadState const& state, [[maybe_unused]] monodroid_should_register should_register) noexcept
+{
+#if defined (DEBUG)
+	const char *last_slash = utils.find_last (entry_name, '/');
+	bool entry_is_overridden = last_slash == nullptr ? false : !should_register (last_slash + 1);
+#else
+	constexpr bool entry_is_overridden = false;
+#endif
+
+	if (register_debug_symbols && !entry_is_overridden && utils.ends_with (entry_name, SharedConstants::PDB_EXTENSION)) {
+		if (bundled_debug_data == nullptr) {
+			bundled_debug_data = new std::vector<XamarinAndroidBundledAssembly> ();
+			bundled_debug_data->reserve (application_config.number_of_assemblies_in_apk);
+		}
+
+		bundled_debug_data->emplace_back ();
+		set_debug_entry_data (bundled_debug_data->back (), state, entry_name);
+		return;
+	}
+
+	if (!utils.ends_with (entry_name, SharedConstants::DLL_EXTENSION)) {
+		return;
+	}
+
+#if defined (DEBUG)
+	if (entry_is_overridden) {
+		return;
+	}
+#endif
+
+	if (bundled_assembly_index >= application_config.number_of_assemblies_in_apk || state.bundled_assemblies_slow_path) [[unlikely]] {
+		if (!state.bundled_assemblies_slow_path && bundled_assembly_index == application_config.number_of_assemblies_in_apk) {
+			log_warn (LOG_ASSEMBLY, "Number of assemblies stored at build time (%u) was incorrect, switching to slow bundling path.", application_config.number_of_assemblies_in_apk);
```

Comment:
Should this be an assert of some form?  Plus other asserts at e.g. startup (somehow)?

Reply:
No, it's just a warning that things are becoming slower than necessary.  We can recover from the condition, no need to abort.